### ISSUE TITLE
Dealing with mixed content and migrating legacy content

### ISSUE THREAD

**Issue by konklone** [GSA/https#11]
In the last day or two, there have been some great threads on the WebAppSec list about making it more straightforward for websites to migrate to HTTPS without having to go and clean up all the legacy URLs (and without disabling mixed content blocking).

* Thread index: https://lists.w3.org/Archives/Public/public-webappsec/2015Feb/thread.html
* Initial `TLS-Only` HTTP header proposal: https://lists.w3.org/Archives/Public/public-webappsec/2015Feb/0006.html
* Proposal  to implement as a CSP directive: https://w3c.github.io/webappsec/specs/upgrade/
* Peter Eckersley of EFF, with 4 proposals of how either browsers or Let's Encrypt could work around the issue: https://lists.w3.org/Archives/Public/public-webappsec/2015Feb/0023.html

The principal issue seems to be: how do you pull this off in such a way that doesn't break your site for older clients that don't support whatever client extension is being proposed to solve it?

**Comment by alex** [comment]
There should probably be more documentation on using CSP's `report-uri` to track mixed content.

**Comment by konklone** [comment]
Also, a tool I found today to help with this: https://github.com/bramus/mixed-content-scan A little hard to install for me, as it's PHP-based, but looks very promising.

I could imagine a web-based tool that performed this scan for your site on demand, and produced a report.

Also, is it possible to provide a CSP reporting endpoint as a service? Or does it have to be on the same domain as the reporting page? (I'm very much a CSP newb.)

**Comment by alex** [comment]
`report-uri` doesn't need to be on the same domain, you could definitely
make a CSP-report-uri-as-a-service. Sentry is looking at adding a CSP
report-uri endpoint: https://github.com/getsentry/sentry/issues/729

On Tue, Feb 3, 2015 at 10:12 PM, Eric Mill <notifications@github.com> wrote:

> Also, a tool I found today to help with this:
> https://github.com/bramus/mixed-content-scan A little hard to install for
> me, as it's PHP-based, but looks very promising.
>
> I could imagine a web-based tool that performed this scan for your site on
> demand, and produced a report.
>
> Also, is it possible to provide a CSP reporting endpoint as a service? Or
> does it have to be on the same domain as the reporting page? (I'm very much
> a CSP newb.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GSA/https/issues/11#issuecomment-72796330>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by konklone** [comment]
I wonder if we could set up a small app that *only* listened for mixed-content reports from `.gov` domains, and filtered out all other requests. The number of `.gov` domains that had this turned on at any given time would probably be manageable, if we were smart about the infrastructure.

**Comment by konklone** [comment]
From Mike West:

> This is something that can certainly be done now. That is, sending `Content-Security-Policy-Report-Only: default-src https:; report-uri /reporting/endpoint` will flood you with the insecure requests your site is making, regardless of whether the site itself is delivered over HTTP/HTTPS. CSP checks happen before mixed content checks for exactly this use case.

From Jacob Hoffman-Andrews, in reply:

> FWIW, this is exactly how we detected mixed content at Twitter, and it worked great. I wrote this up on Stack Overflow ages ago, but I've been meaning to write a proper blog post about it: http://stackoverflow.com/questions/4129496/detect-broken-lock-icon-mixed-secure-insecure-content-from-javascript.
> 
> What we did in practice: The CSP header for mixed content detection was sent only on HTTPS requests and only to 1% of live users, to avoid flooding the endpoint. Also, the same header was set in enforce mode for 100% of employee accounts, so engineers who introduced mixed content would see it right away (this was before mixed content blocking, though it continued to be useful for mixed passive content).

**Comment by konklone** [comment]
I'v been playing around with [`mixed-content-scan`](https://github.com/bramus/mixed-content-scan) and it's pretty good. I [added a JSON output option](https://github.com/bramus/mixed-content-scan/pull/16) so that we could more easily wire it into a reporting pipeline (dedupe on assets, etc.).

Given a domain or subdomain, this could crawl through a site based on seen links (which is potentially incomplete -- I'd appreciate being able to [use a sitemap](https://github.com/bramus/mixed-content-scan/issues/15)) -- and produce a set of unique URLs and domains that need to be addressed, and what pages produced them. A bit of extra tooling could then probe each found URL to see whether a valid HTTPS version of the URL already exists.

**Comment by konklone** [comment]
@tollmanz points me to [a Wordpress plugin](https://wordpress.org/plugins/https-mixed-content-detector/screenshots/) that uses CSP reporting for mixed content detection.

**Comment by tollmanz** [comment]
I'm just jumping into this thread not knowing a thing about this project, but wanted to mention a few things about the [HTTPS Mixed Content Detector WordPress plugin](https://github.com/tollmanz/wordpress-https-mixed-content-detector) (Github link). I doubt you are on WP, but the mechanisms for what you are doing are all available in this plugin and I thought my experiences building it might be useful for this discussion.

The plugin sets a very specific CSP report only policy. Unfortunately, I'm not at my dev machine so I cannot show the header version of the policy, but as a PHP array, it looks like this:

```php
array(
	'default-src' => "https: data: 'unsafe-inline' 'unsafe-eval'",
	'child-src'   => "https:",
	'connect-src' => "https:",
	'font-src'    => "https: data:",
	'img-src'     => "https: data:",
	'media-src'   => "https:",
	'object-src'  => "https:",
	'script-src'  => "https: 'unsafe-inline' 'unsafe-eval'",
	'style-src'   => "https: 'unsafe-inline'",
);
```

The plugin uses a very specific CSP in order to only alert on mixed content issues and to *group violations by type*. This helps you solve certain types of violations differently than others. For instance, all image violations might be resolved by changing the URLs to point at a newly configured TLS enabled CDN, whereas violating scripts may need to be handled differently.

The plugin is set up to run on HTTP, yet report HTTPS mixed content issues to the beacon. Since this is in a WordPress context (and WordPress folks loving owning their own content), the beacon is sent to a URL on the site owner's website and all data is collected there. Because you can run it on HTTP, you are able to begin fixing content issues *before* you flip the switch to go all HTTPS...pretty handy.

The plugin uses cURL (when available) to detect if a secure variant of the content is available (i.e., can we connect to the `https` version of the content). This allows you to know whether or not you can just swap out the `http` variant for the `https` variant. Since its WordPress, I'm looking to then build filters that will swap this out automatically so the mixed content issue should only appear once.

The plugin has two modes, *admin only* and *sampling*. In admin only mode, the CSP report only header is only set for admins, meaning that mixed content warnings will only be triggered for those individuals. This is helpful as it reduces the number of warnings (helps with flooding) and allows for a authentication mechanism with the beacon. Note that I learned that Chrome and Safari will pass the cookie to the report only URI, whereas Firefox will not meaning that this mode is broken for Firefox because the cookie-based auth is broken.

Sampling mode opens up reporting for every user, but as the name suggests, it samples traffic. You can dial in a level of precision to determine how many of the reports the beacon will accept. The advantage of this mode is that you get real world traffic sampling from *all browsers that support CSP*. The disadvantage is that you are allowing unauthenticated requests to the beacon and you are leaving yourself wide open for an easy DoS-ing. Obviously, I do not recommend leaving the sampling mode on indefinitely.

I hope this info is helpful. While your context is not WordPress, the general idea applies to any web deployment so I thought I'd share a bit.

**Comment by konklone** [comment]
This is *extremely* helpful, thank you very much for the info, @tollmanz. And I know there are many Wordpress sites in the .gov space, so it's very relevant.

**Comment by jsha** [comment]
At Twitter, our collection endpoint just stored records for offline analysis, so we weren't worried about being flooded with unauthenticated requests as a DoS. It's really nice to be able to keep your collection headers in sampling mode permanently, since there will be regressions over time, unless you turn on enforce mode, which is ideal. One of our constraints at the time was that we had third-party passive mixed content that we couldn't fix, and the business side was unwilling to give up on integrating. These were, e.g. third party image hosting and video hosting in an iframe. We whitelisted those in the CSP header.

We also found a surprising number of unexpected schemes in our own code, like iframes with src="javascript:...". We kept tweaking our CSP until it whitelisted all the schemes in all the places we needed.

We got a ton of bogus reports from people with browser extensions that inserted mixed content into the page. These were often hard to distinguish from legitimate reports because those extensions inserted Twitter-relevant content, like profile images. I'm guessing this will be less of a problem with .gov. We also saw weird proxy sites. Some would simply point foo.example.com's DNS at Twitter, and some number of people would visit it, then send CSP reports with a Referrer of foo.example.com. Other weird proxy sites would host plaintext HTTP at bar.example.com, then proxy every request for, e.g. /j4cob to https://twitter.com/j4cob. These would generate a ton of bogus CSP reports. So you probably want to filter by Referrer, like you said.

In general you want to group and count reports by the page or section that generated them, so you can prioritize which ones to fix. Sometimes there will be an important site area that's generating lots of reports, but it gets spread thin across many Referrer URLs because there are many unique URLs based on query parameters or magic path components. You'll want some very simple rules in your logs analysis to group these reports back together usefully. You'll always get some background noise of reports from Internet weirdness, so you're mainly looking for stuff that shows up in sufficient volume.

Ideally you want your nightly logs analysis script to detect when any given page generates more than N mixed content reports, so you can fix it right away.

**Comment by konklone** [comment]
Another piece on Twitter's CSP work, [written by Neil Matatall](https://twitter.com/ndm/status/565593697002455040):

https://blog.matatall.com/2014/07/twitters-csp-report-collector-design/

**Comment by alex** [comment]
For folks who don't have a PHP environment set up but want to run `mixed-content-scan`, you can throw these files in a directory (https://gist.github.com/anonymous/914944cc4542425539b5) and run `fig run scan` (you'll need `fig` and `docker` installed), and then in the bash shell `mixed-content-scan https://url`.

**Comment by konklone** [comment]
Re-opening to make sure we integrate some CSP auto-reporting guidance.

**Comment by cherrypj** [comment]
What about Protocol Relative URLs (or just relative URLs)? You know, those ones that start with //, like this:

    <img src="//example.com/image.jpg" />

Those are our friends. Won't they load that resource with whatever protocol the current page is? 

Links that are just relative to begin with will also be OK:

    <img src="/images/image.jpg" />

**Comment by konklone** [comment]
Closing, as it's almost all done now, and will be resolved with #117.
